### PR TITLE
Added moshi and started to convert data model to kotlin

### PR DESCRIPTION
**Pull request by tomaszpolanski**

### CODE REVIEW

File: app/src/test/java/com/futurice/freesound/test/data/TestData.kt [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            return SoundSearchResult.builder()
+                    .count(count * 2)
+                    .next("nextUrl")
+                    .previous("prevUrl")
+                    .results(sounds(count))
+                    .build()
+        }
+
+        fun sounds(count: Int): List<Sound> {
+            return (1L..(count - 1))
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/test/data/TestData.kt]: Is there a nicer way to do an exclusive range?

File: app/src/test/java/com/futurice/freesound/test/data/TestData.kt [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                    .map { sound(it) }
+                    .toList()
+        }
+
+        fun sound(index: Long?): Sound {
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/test/data/TestData.kt]: Shouldn't be nullable

File: app/src/test/java/com/futurice/freesound/test/data/TestData.kt [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            return Sound.builder()
+                    .id(index!!)
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/test/data/TestData.kt]: Then no need for `!!`

File: app/src/test/java/com/futurice/freesound/test/data/TestData.kt [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                    .url(indexed("url", index))
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/test/data/TestData.kt]: `indexed` function could probably just become "url $index". etc elsewhere.

File: app/build.gradle [hunk 5]
```diff
@@ -155,6 +156,11 @@ dependencies {
     implementation "com.jakewharton.retrofit:retrofit2-rxjava2-adapter:$dependencyVersions.retrofit2RxJava2Adapter"
     implementation "com.squareup.retrofit2:converter-gson:$dependencyVersions.retrofit2GsonConverter"
 
+    // Moshi
+    implementation "com.squareup.moshi:moshi:$dependencyVersions.moshi"
+    implementation "com.squareup.moshi:moshi-kotlin:$dependencyVersions.moshi"
+    implementation "com.squareup.retrofit2:converter-moshi:$dependencyVersions.moshiConverter"
+
     // GSON
     implementation "com.google.code.gson:gson:$dependencyVersions.gson"
```
Comment (tomaszpolanski) [inline app/build.gradle]: Yes, didn't yet replace all - too big PR

File: app/src/main/java/com/futurice/freesound/network/api/model/AccessToken.kt [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package com.futurice.freesound.network.api.model
+
+import com.squareup.moshi.Json
+
+data class AccessToken(
+        @Json(name = "access_token") val accessToken: String,
+        val scope: String,
+        @Json(name = "expires_in") val expiresIn: Long = 0L,
```
Comment (tomaszpolanski) [inline app/src/main/java/com/futurice/freesound/network/api/model/AccessToken.kt]: I can make it nullable, simply it was too much of safeguarding

File: app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt [hunk 7]
```diff
@@ -0,0 +1,419 @@
+/*
```
Comment (tomaszpolanski) [inline app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt]: Before we had to have a builder for that, now we don't - so it's smaller

File: app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+    fun userAvatar_isUserMediumAvatar() {
+        val username = "username"
+        val avatar_m = "avatar_m"
+        val sound = TEST_SOUND.toBuilder()
+                .username(username)
+                .build()
+        val user = TestData.user()
+                .copy(avatar = TestData.avatar().copy(medium = avatar_m))
+
+        ArrangeBuilder().withUserResponse(username, user)
```
Comment (tomaszpolanski) [inline app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt]: Yes, didn't want to have it in this PR

File: app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+                .assertValue(avatar_m)
+    }
+
+    @Test
+    fun username_isUsersUsername() {
+        val username = "username"
+        val sound = TEST_SOUND.toBuilder()
+                .username(username)
+                .build()
+        val user = TestData.user().copy(username = username)
```
Comment (tomaszpolanski) [inline app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt]: This is not really `this.username` as this is not class field

File: app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+    @Test
+    fun thumbnailImageUrl_emitsSoundMediumWaveformUrl() {
+        val soundItemViewModel = SoundItemViewModel(TEST_SOUND,
+                navigator,
+                audioPlayer,
+                freeSoundApiService)
+
+        soundItemViewModel.thumbnailImageUrl()
+                .test()
+                .assertValue(TEST_SOUND.images()!!.medSizeWaveformUrl())
```
Comment (tomaszpolanski) [inline app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt]: I know

File: app/src/test/java/com/futurice/freesound/test/data/TestData.kt (hunk 1 above)
Comment (tomaszpolanski) [inline app/src/test/java/com/futurice/freesound/test/data/TestData.kt]: Apart for having `0..count` and that adding 1, no

Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/test/data/TestData.kt]: Ok. Let just make it `0..count` (and also further down too). No tests should be relying on the specific index values.

File: app/src/main/java/com/futurice/freesound/network/api/model/AccessToken.kt (hunk 6 above)
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/network/api/model/AccessToken.kt]: Let's not have a default and leave it non-null.

File: app/src/test/java/com/futurice/freesound/feature/home/DefaultUserDataModelTest.kt [hunk 11]
```diff
@@ -35,6 +35,8 @@ class DefaultUserDataModelTest {
 
     private lateinit var dataModel: DefaultUserDataModel
 
+    private val user: User get() = TestData.user()
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/home/DefaultUserDataModelTest.kt]: let's name it `testUser`, I think `user { user }` is a little confusing.

File: app/src/test/java/com/futurice/freesound/feature/home/HomeFragmentViewModelTest.kt [hunk 12]
```diff
@@ -17,11 +17,12 @@ class HomeFragmentViewModelTest {
     @Mock
     private lateinit var userDataModel: UserDataModel
 
-    private val USER = TestData.user()
+    private lateinit var USER: User
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/home/HomeFragmentViewModelTest.kt]: Take the same approach as the other class.

File: app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt (hunk 7 above)
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt]: I would have thought it was much smaller. Oh well.

File: app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt (hunk 9 above)
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.kt]: Yep. I misread it.

**Pull request merged**